src/msg/async/Stack.h

```diff
@@ -0,0 +1,315 @@
+  ///          a \ref entity_addr_t describing the remote endpoint.
+  int accept(ConnectedSocket *sock, const SocketOptions &opt, entity_addr_t *out) {
+    return _ssi->accept(sock, opt, out);
+  }
+
+  /// Stops any \ref accept() in progress.
+  ///
+  /// Current and future \ref accept() calls will terminate immediately
+  /// with an error.
+  void abort_accept() {
+    _ssi->abort_accept();
+    _ssi.reset();
+  }
+
+  /// Get file descriptor
+  int fd() const {
+    return _ssi->fd();
+  }
+
+  explicit operator bool() const {
+    return _ssi.get();
+  }
+};
+/// @}
+
+class NetworkStack;
+
+enum {
+  l_msgr_first = 94000,
+  l_msgr_recv_messages,
+  l_msgr_send_messages,
+  l_msgr_send_messages_inline,
+  l_msgr_recv_bytes,
+  l_msgr_send_bytes,
+  l_msgr_created_connections,
+  l_msgr_active_connections,
+  l_msgr_last,
+};
+
+class Worker {
+  std::mutex init_lock;
+  std::condition_variable init_cond;
+  bool init = false;
+
+ public:
+  bool done = false;
+
+  CephContext *cct;
+  PerfCounters *perf_logger;
+  unsigned id;
+
+  std::atomic_uint references;
+  EventCenter center;
+  Worker(CephContext *c, unsigned i)
+    : cct(c), perf_logger(NULL), id(i), references(0), center(c) {
+    char name[128];
+    sprintf(name, "AsyncMessenger::Worker-%d", id);
+    // initialize perf_logger
+    PerfCountersBuilder plb(cct, name, l_msgr_first, l_msgr_last);
+
+    plb.add_u64_counter(l_msgr_recv_messages, "msgr_recv_messages", "Network received messages");
+    plb.add_u64_counter(l_msgr_send_messages, "msgr_send_messages", "Network sent messages");
+    plb.add_u64_counter(l_msgr_send_messages_inline, "msgr_send_messages_inline", "Network sent inline messages");
+    plb.add_u64_counter(l_msgr_recv_bytes, "msgr_recv_bytes", "Network received bytes");
+    plb.add_u64_counter(l_msgr_send_bytes, "msgr_send_bytes", "Network received bytes");
+    plb.add_u64_counter(l_msgr_active_connections, "msgr_active_connections", "Active connection number");
+    plb.add_u64_counter(l_msgr_created_connections, "msgr_created_connections", "Created connection number");
+
+    perf_logger = plb.create_perf_counters();
+    cct->get_perfcounters_collection()->add(perf_logger);
+  }
+  virtual ~Worker() {
+    if (perf_logger) {
+      cct->get_perfcounters_collection()->remove(perf_logger);
+      delete perf_logger;
+    }
+  }
+
+  virtual int listen(entity_addr_t &addr,
+                     const SocketOptions &opts, ServerSocket *) = 0;
+  virtual int connect(const entity_addr_t &addr,
+                      const SocketOptions &opts, ConnectedSocket *socket) = 0;
+
+  virtual void initialize() {}
+  PerfCounters *get_perf_counter() { return perf_logger; }
+  void release_worker() {
+    int oldref = references.fetch_sub(1);
+    assert(oldref > 0);
+  }
+  void init_done() {
+    init_lock.lock();
+    init = true;
+    init_cond.notify_all();
+    init_lock.unlock();
+  }
+  void wait_for_init() {
+    std::unique_lock<std::mutex> l(init_lock);
```

Comment:
I think we must use unique_lock, otherwise:
msg/async/Stack.h:261:23: error: no matching function for call to
'std::condition_variable::wait(std::lock_guard<std::mutex>&)'
       init_cond.wait(l);


On Tue, Aug 2, 2016 at 11:04 PM, Kefu Chai <notifications@github.com> wrote:

> In src/msg/async/Stack.h
> <https://github.com/ceph/ceph/pull/10264#discussion_r73172734>:
>
> > +                      const SocketOptions &opts, ConnectedSocket *socket) = 0;
> > +
> > +  virtual void initialize() {}
> > +  PerfCounters *get_perf_counter() { return perf_logger; }
> > +  void release_worker() {
> > +    int oldref = references.fetch_sub(1);
> > +    assert(oldref > 0);
> > +  }
> > +  void init_done() {
> > +    init_lock.lock();
> > +    init = true;
> > +    init_cond.notify_all();
> > +    init_lock.unlock();
> > +  }
> > +  void wait_for_init() {
> > +    std::unique_lock<std::mutex> l(init_lock);
>
> @yuyuyu101 <https://github.com/yuyuyu101> ping?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ceph/ceph/pull/10264/files/4a367d4229270b6a21059855d19d4bacc79b46bc#r73172734>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA72YY12Rfc_H3mxGBXsI_iYnlnlbmigks5qb1xugaJpZM4JKlyl>
> .
>



-- 

Best Regards,

Wheat